File: app/src/main/java/com/domclick/controller/AccountController.java
package com.domclick.controller;

import com.domclick.model.Account;
import com.domclick.repository.AccountRepository;
import com.domclick.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.validation.BindingResult;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;

import javax.validation.Valid;

import static java.lang.String.format;

@Controller
public class AccountController {
    @Autowired
    private UserRepository userRepository;

    @Autowired
    private AccountRepository accountRepository;

    @GetMapping(value = "/accounts")
    public String accountsList(Model model) {
        model.addAttribute("accounts", accountRepository.findAll());
        return "account/list";
    }

    @GetMapping(value = {"/accounts/edit", "/accounts/edit/{id}"})
    public String accountEditForm(Model model, @PathVariable(required = false, name = "id") Long id) {
        Account account = id != null ? findAccountById(id) : new Account();
        account.updateUserId();
        model.addAttribute("account", account);

        model.addAttribute("users", userRepository.findAll());
        return "account/edit";
    }

    @PostMapping(value = "/accounts/edit")
    public String doAccountEdit(Model model, @Valid Account account, BindingResult bindingResult) {
        if (bindingResult.hasErrors()) {
            return "account/edit";
        }

        Account reload = account.getId() != null ? findAccountById(account.getId()) : null;
        if (reload != null) {
            reload.setBalance(account.getBalance());
            reload.setUser(userRepository.findById(account.getUserId()).orElseThrow(() -> new RuntimeException(
                    format("Can not found account by id '%s'", account.getUserId()))));
            accountRepository.save(reload);
        } else {
            account.setUser(userRepository.findById(account.getUserId()).orElseThrow(() -> new RuntimeException(
                    format("Can not found account by id '%s'", account.getUserId()))));
            accountRepository.save(account);
        }
        model.addAttribute("accounts", accountRepository.findAll());
        return "account/list";
    }

    @GetMapping(value = "/accounts/delete/{id}")
    public String accountDelete(Model model, @PathVariable(name = "id") Long id) {
        accountRepository.delete(findAccountById(id));
        model.addAttribute("accounts", accountRepository.findAll());
        return "account/list";
    }

    private Account findAccountById(Long id) {
        return accountRepository.findById(id).orElseThrow(() -> new RuntimeException(
                format("Can not found account by id '%s'", id)));
    }
}

File: app/src/main/kotlin/com/domclick/controller/AccountController.kt
package com.domclick.controller

import com.domclick.entity.AccountEntity
import com.domclick.exception.BadRequestException
import com.domclick.service.AccountService
import com.domclick.service.UserService
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.validation.BindingResult
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import java.lang.String.format
import javax.validation.Valid

@Controller
class AccountController(
        private val userService: UserService,
        private val accountService: AccountService
) {

    @GetMapping(value = ["/accounts"])
    fun accountsList(model: Model): String {
        model.addAttribute("accounts", accountService.findAll())
        return "account/list"
    }

    @GetMapping(value = ["/accounts/edit", "/accounts/edit/{id}"])
    fun accountEditForm(model: Model, @PathVariable(required = false, name = "id") id: Long?): String {
        val account = if (id != null) findAccountById(id) else AccountEntity()

        model.addAttribute("account", account)
        model.addAttribute("users", userService.findAll())
        if (account.user != null) account.updateUserId()
        return "account/edit"
    }

    @PostMapping(value = ["/accounts/edit"])
    fun doAccountEdit(model: Model, @Valid account: AccountEntity, bindingResult: BindingResult): String {
        if (bindingResult.hasErrors()) {
            return "account/edit"
        }

        accountService.upsert(account)

        model.addAttribute("accounts", accountService.findAll())
        return "account/list"
    }

    @GetMapping(value = ["/accounts/delete/{id}"])
    fun accountDelete(model: Model, @PathVariable(name = "id") id: Long): String {
        val reload = findAccountById(id) ?: return "account/list"
        accountService.delete(reload)
        model.addAttribute("accounts", accountService.findAll())
        return "account/list"
    }

    private fun findAccountById(id: Long) =
            accountService.findById(id).orElseThrow {
                throw BadRequestException(format("AccountEntity with id '%s' not found", id))
            }
}
